### PR TITLE
Introduce reset mechanics to microcontroller.

### LINKED ISSUE #173: Add ability to reset into safe mode from the REPL
If no file system is writeable from the host computer do it automatically. If it is writable then require force=True kwarg.

### LINKED ISSUE #350: No easy way to reboot into safe mode on the Feather M0 Express
While working on issue #286, I discovered that there is not an easy way (say shorting a pin to GND or 3V while pushing reset) to reboot the Feather M0 Express into safe mode. I messed up some test scripts and ended up getting the `CIRCUITPY` drive stuck in read only mode (read only from the PC) and had to temporarily flash firmware with safe mode forced on to fix it (this was easier than switching to Arduino and wiping the SPI flash).

The Feather M0 Express doesn't have a switchable safe mode implemented in `atmel-samd/boards/feather_m0_express/board.c`:

```c
bool board_requests_safe_mode(void) {
    return false;
}
```

If this is something that you would like done, I'd be happy to work on it.

### PR DESCRIPTION
**Pull request by tannewt**
This allows one to configure how a subsequent reset will behave and
also trigger a reset.

Fixes #350 and fixes #173

### CODE REVIEW

File: shared-bindings/microcontroller/RunMode.c [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ */
+
+#include "shared-bindings/microcontroller/RunMode.h"
+
+//| .. currentmodule:: microcontroller
+//|
+//| :class:`RunMode` -- run state of the microcontroller
+//| =============================================================
+//|
+//| .. class:: mcirocontroller.RunMode
```
Comment (dhalbert) [inline shared-bindings/microcontroller/RunMode.c]: spelling error

File: atmel-samd/main.c [hunk 2]
```diff
@@ -524,10 +528,14 @@ safe_mode_t samd21_init(void) {
...
+    // survives we enter safe mode with a friendlier message.
+    bool software_safe_mode = _ezero == CIRCUITPY_SOFTWARE_SAFE_MODE;
+
 // On power on start or external reset, set _ezero to the canary word. If it
-// gets killed, we boot in safe mod. _ezero is the boundary between statically
-// allocated memory including the fixed MicroPython heap and the stack. If either
-// misbehaves, the canary will not be in tact after soft reset.
+// gets killed, we boot in safe mode. _ezero is the boundary between statically
+// allocated memory including the fixed MicroPython heap and the stack. If
+// either misbehaves, the canary will not be in tact after soft reset.
```
Comment (dhalbert) [inline atmel-samd/main.c]: "in tact" -> "intact"

**Comment by dhalbert** [comment]
Here's an example of a use case: I have something running a CPy program, and I want to do a simple reset to have it restart. So I need to press reset twice? I don't want the user to know about the bootloader, ideally.

For instance, the assistive technology CPy-based button box I built has a big red reset button in case it gets confused. But now it has to be pressed twice?

**Comment by tannewt** [comment]
I'm not sure what you mean by simple reset to restart. There is a separate issue (#184) for being able to reload code but not reset the whole microcontroller.

The default behavior stays the same. One reset press will reset and load the user code (circuitpython) and two quick presses will stay in the bootloader.

This code adds the ability to trigger the bootloader on the next reset so no double tap is needed. One can also trigger safe mode this way by setting safe mode as the next target and triggering a reset.

**Comment by dhalbert** [comment]
Ah, OK! I thought it was going into the bootloader by default after one click. I missed the last sentence [here](https://github.com/adafruit/circuitpython/issues/350#issuecomment-349150782):
>Adding the ability to reset to the bootloader next reset press means we can provide the same out-of-box experience to the CircuitPlayground Express even though it'll ship with CircuitPython going forwards. We'll ship with demo code that make subsequent resets go to the bootloader rather than requiring a double tap. The double tap behavior will happen once that demo code is replaced though.

Sorry for any crankiness. I should have asked earlier.

**Comment by dhalbert** [comment]
oops, forgot to wait for spelling corrections

**Comment by tannewt** [comment]
Just committed the spelling fixes.

**Pull request merged**